iocore/net/quic/qlog/QLogFrame.cc

```diff
@@ -0,0 +1,282 @@
+/** @file
+ *
+ *  A brief file description
+ *
+ *  @section license License
+ *
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "QLogFrame.h"
+
+namespace QLog
+{
+template <typename Real>
+Real &
+Convert(const QUICFrame *frame)
+{
+  // FIXME: dangerous
+  auto tmp = const_cast<QUICFrame *>(frame);
```

Comment:
I don't understand why we need to remove `const`. I haven't compiled the new code yet so I'm not sure if it's possible with the new code, but we should keep `const`.

Reply:
```cpp
template <typename Real>                                                            
const Real &                                                                        
Convert(const QUICFrame *frame)                                                     
{                                                                                   
#if defined(DEBUG)                                                                  
  auto ref = dynamic_cast<const Real *>(frame);                                     
  ink_assert(ref != nullptr);                                                       
  return *ref;                                                                      
#endif                                                                              
  return *static_cast<const Real *>(frame);                                                                                                                                      
} 
```
```cpp
return std::make_unique<Frame::AckFrame>(Convert<const QUICAckFrame>(frame));
```
```cpp
AckFrame(const QUICAckFrame &frame) : QLogFrame(frame.type())
```